subsys/bluetooth/host/classic/avdtp.c

```diff
@@ -299,61 +363,77 @@ static void avdtp_get_capabilities_handler(struct bt_avdtp *session, struct net_
 	} else {
 		struct bt_avdtp_req *req = session->req;
 
-		if (session->req == NULL) {
+		if (req == NULL) {
 			return;
 		}
 		k_work_cancel_delayable(&session->timeout_work);
 		GET_CAP_REQ(session->req)->buf = NULL;
 
 		if (msg_type == BT_AVDTP_ACCEPT) {
-			GET_CAP_REQ(session->req)->status = 0;
-			if (session->req != NULL) {
-				GET_CAP_REQ(session->req)->buf = buf;
-			}
-		} else if (msg_type == BT_AVDTP_REJECT) {
-			GET_CAP_REQ(session->req)->status = net_buf_pull_u8(buf);
-		} else if (msg_type == BT_AVDTP_GEN_REJECT) {
-			GET_CAP_REQ(session->req)->status = BT_AVDTP_NOT_SUPPORTED_COMMAND;
+			GET_CAP_REQ(req)->buf = buf;
 		}
-		AVDTP_LOCK();
-		session->req = NULL;
-		AVDTP_UNLOCK();
+		avdtp_set_status(req, buf, msg_type);
+		bt_avdtp_clear_req(session);
 		if (req->func != NULL) {
 			req->func(req);
 		}
 	}
 }
 
 static void avdtp_process_configuration(struct bt_avdtp *session, struct net_buf *buf,
-					uint8_t msg_type, uint8_t tid)
+					uint8_t msg_type, uint8_t tid, bool reconfig)
 {
 	if (msg_type == BT_AVDTP_CMD) {
 		int err = 0;
+		int ret;
 		struct bt_avdtp_sep *sep;
 		struct net_buf *rsp_buf;
 		uint8_t error_code = 0;
 
-		sep = avdtp_get_sep(net_buf_pull_u8(buf) >> 2);
-		if ((sep == NULL) || (session->ops->set_configuration_ind == NULL)) {
+		sep = avdtp_get_cmd_sep(buf);
+		if (sep == NULL) {
+			err = -ENOTSUP;
+		} else if (!reconfig && session->ops->set_configuration_ind == NULL) {
+			err = -ENOTSUP;
+		} else if (reconfig && session->ops->re_configuration_ind == NULL) {
 			err = -ENOTSUP;
 		} else {
-			if (sep->state == AVDTP_STREAMING) {
+			uint8_t expected_state;
+
+			if (reconfig) {
+				expected_state = AVDTP_OPEN | AVDTP_OPENING;
+			} else {
+				expected_state = AVDTP_IDLE;
+			}
+
+			avdtp_sep_lock(sep);
```

Comment:
Hierarchically, this `lock` has no corresponding `unlock`.

Reply:
Fixed